File: src/main/java/ru/ifmo/ctddev/segal/hw3/RungeKuttaMethod.java
package ru.ifmo.ctddev.segal.hw3;

import ru.ifmo.ctddev.segal.hw3.niftystuff.Point;
import ru.ifmo.ctddev.segal.hw3.niftystuff.TriDimensionalFunction;

/**
 * @author Ignat Loskutov
 */
public class RungeKuttaMethod extends MethodForLorenzSystem {
    /**
     * Instantiates Runge—Kutta fourth-order method.
     *
     * See: https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods.
     */
    public RungeKuttaMethod(double sigma, double b, double r, double ta, double tb, double dt, double x0, double y0, double z0) {
        super(sigma, b, r, ta, tb, dt, x0, y0, z0);
    }

    @Override
    public Result call() {
        ResultBuilder builder = new ResultBuilder();

        TriDimensionalFunction f = p -> new Point(firstEquation(p.x, p.y, p.z),
                                                  secondEquation(p.x, p.y, p.z),
                                                  thirdEquation(p.x, p.y, p.z));

        Point p = new Point(x0, y0, z0);
        for (double t = ta; t <= tb; t += dt) {
            builder.append(p.x, p.y, p.z, t);

            Point k1 = f.apply(p);
            Point k2 = f.apply(p.add(k1.mul(dt/2)));  // fuck Java :/
            Point k3 = f.apply(p.add(k2.mul(dt/2)));
            Point k4 = f.apply(p.add(k3.mul(dt)));

            p = p.add(k1.add(k2.mul(2)).add(k3.mul(2)).add(k4).mul(dt/6));
        }
        return builder.toResult();
    }
}

File: src/main/java/ru/ifmo/ctddev/segal/hw3/RungeKuttaMethod.kt
package ru.ifmo.ctddev.segal.hw3

import ru.ifmo.ctddev.segal.hw3.niftystuff.Point

/**
 * Implements Runge—Kutta fourth-order method.
 * @author Ignat Loskutov
 */

class RungeKuttaMethod(sigma: Double, b: Double, r: Double,
                       ta: Double, tb: Double, dt: Double,
                       x0: Double, y0: Double, z0: Double)
: MethodForLorenzSystem(sigma, b, r, ta, tb, dt, x0, y0, z0) {
    override fun call(): Result {
        val builder = ResultBuilder()
        val f = { p: Point -> Point(firstEquation(p.x, p.y, p.z),
                                    secondEquation(p.x, p.y, p.z),
                                    thirdEquation(p.x, p.y, p.z)) }
        var p = Point(x0, y0, z0)
        var t = ta
        while (t <= tb) {
            builder.append(p.x, p.y, p.z, t)

            val k1 = f(p)
            val k2 = f(p + k1 * (dt / 2))
            val k3 = f(p + k2 * (dt / 2))
            val k4 = f(p + k3 * dt)

            p += (k1 + k2 * 2.0 + k3 * 2.0 + k4) * (dt / 6)
            t += dt
        }
        return builder.toResult()
    }
}


